### ISSUE TITLE
Feature request:  Built in i18n support

### ISSUE THREAD

**Issue by rottenpen** [denoland/fresh#666]
I want to be responsible for i18n supporting, can you be my mentor?

this is my tiny rfc for i18n : https://github.com/rottenpen/fresh/commit/e208eef77a40bb00543a308b7e0bc403968daa56

- This case ⬆️ has the internationalized language switch between zh-CN and en-US
- use i18n-runtime to get the i18n content
- i18n.config.js is the options for fresh/plugins/i18n.ts

**Comment by ch99q** [comment]
It's possible to do this already using the new plugin system, but if you wish to handle complex language switching and cross-domain language switch, the plugin system needs to be extended to support root middleware injection.

I have a current implementation that allows for complex i18n localization with fluentjs as a message provider. To handle the logic I had to apply some context on the req before `ctx.next()` and then apply some logic / replace responses with a redirect in some cases.

I'm working on separating it into a simple module, but as far as I can see, if the plugin system just supports as `middleware` function, it's possible to make very complex solutions with minimal code.

**Comment by likonefokou** [comment]
any update on this topic ? is i18n effective in fresh ?

**Comment by perguth** [comment]
One of the OPs in Discord asked for feedback on how people are doing i18n here: https://discord.com/channels/684898665143206084/1193992623413792819/1194028132039131206

He said the general approach at the moment is:

![image](https://github.com/user-attachments/assets/b52b2421-6664-401d-8005-6b38a78ac1a4)

**Comment by KishiTheMechanic** [comment]
I recently created an i18n plugin called fresh-i18n using the Fresh v2 plugin interface, and it's now available on the jsr module registry.

JSR: https://jsr.io/@elsoul/fresh-i18n
GitHub: https://github.com/elsoul/fresh-i18n

I've also set up a demo site, so feel free to explore it and give the plugin a try.

Demo: https://fresh-web.skeeter.dev/
GitHub: https://github.com/elsoul/skeet-v3-fresh-edge-web-rc

Any feedback is always welcome. 
Happy coding👨‍💻 LFG🚀🔥

**Comment by perguth** [comment]
@KishiTheMechanic: I do not yet see how this can be the way to go:

It looks like for Islands your lib expects that you send out the whole translation object (the whole context.state in the example) for every Island you include in the page that uses translations: https://github.com/elsoul/fresh-i18n?tab=readme-ov-file#step-4-use-translation-in-islands

Or do I misunderstand?

**Comment by KishiTheMechanic** [comment]
@perguth Yes, it will share the translation object when needed but the object should have only partial data. (The retrieve json data will be decided from path name. eg. when the access is to /company/about, common.json (also error.json and metadata.json), company.json and about.json will be read if they exist.

I also created fresh-atom, which creates a global store using Preact's signals. This allows data to be shared across islands.

https://jsr.io/@elsoul/fresh-atom
